src/libsystemd/sd-device/device-trigger.c

```diff
@@ -0,0 +1,483 @@
+        assert_return_errno(trigger, NULL, EINVAL);
+
+        return trigger->action;
+}
+
+_public_ const char *sd_device_trigger_get_uuid(sd_device_trigger *trigger) {
+        int r;
+
+        assert_return(trigger, NULL);
+
+        r = device_trigger_ensure_uuid_set(trigger);
+        if (r < 0)
+                return NULL;
+
+        return trigger->uuid;
+}
+
+_public_ int sd_device_trigger_set_source(sd_device_trigger *trigger, const char *source) {
+        assert_return(trigger, -EINVAL);
+
+        if (trigger->executing)
+                return -EPERM;
+
+        if (!in_charset(source, ALPHANUMERICAL))
+                return -EINVAL;
+
+        trigger->source = source;
+
+        return 0;
+}
+
+_public_ const char *sd_device_trigger_get_source(sd_device_trigger *trigger) {
+        assert_return_errno(trigger, NULL, EINVAL);
+
+        return trigger->source;
+}
+
+_public_ int sd_device_trigger_add_device(sd_device_trigger *trigger, sd_device *device) {
+        assert_return(trigger, -EINVAL);
+        assert_return(device, -EINVAL);
+
+        if (trigger->executing)
+                return -EPERM;
+
+        if (!GREEDY_REALLOC(trigger->devices, trigger->n_allocated, trigger->n_devices + 1))
+               return -ENOMEM;
+
+        trigger->devices[trigger->n_devices++] = sd_device_ref(device);
+
+        return 0;
+}
+
+_public_ int sd_device_trigger_add_enumerator(sd_device_trigger *trigger, sd_device_enumerator *enumerator) {
+        sd_device **devices;
+        size_t i, n_devices;
+
+        assert_return(trigger, -EINVAL);
+        assert_return(trigger, -EINVAL);
+
+        if (trigger->executing)
+                return -EPERM;
+
+        devices = device_enumerator_get_devices(enumerator, &n_devices);
+
+        if (!devices || !n_devices)
+                return 0;
+
+        if (!GREEDY_REALLOC(trigger->devices, trigger->n_allocated, trigger->n_devices + n_devices))
+                return -ENOMEM;
+
+        for (i = 0; i < n_devices; i++)
+                trigger->devices[trigger->n_devices++] = sd_device_ref(devices[i]);
+
+        return 0;
+}
+
+sd_device **device_trigger_get_devices(sd_device_trigger *trigger, size_t *ret_n_devices) {
+        assert_return(trigger, NULL);
+        assert_return(ret_n_devices, NULL);
+
+        *ret_n_devices = trigger->n_devices;
+        return trigger->devices;
+}
+
+_public_ int sd_device_trigger_add_property(sd_device_trigger *trigger, const char *_key, const char *_value) {
+        _cleanup_free_ char *key = NULL;
+        _cleanup_free_ char *value = NULL;
+        int r;
+
+        assert_return(trigger, -EINVAL);
+        assert_return(_key, -EINVAL);
+        assert_return(_value, -EINVAL);
+
+        if (trigger->executing)
+                return -EPERM;
+
+        if (!in_charset(_key, ALPHANUMERICAL) ||
+            !in_charset(_value, ALPHANUMERICAL))
+                return log_debug_errno(-EINVAL, "sd-device-trigger: Incorrect property '%s=%s': "
```

Comment:
no need to negate the error code here. log_debug_errno() returns `-abs()` of the specified parameter anyway, so that negative and positive errors are the same